File: experiments/source/cpp/dot_pipeline.cpp
#include <loom/loom.hpp>

LOOM_ENTRY float dot_pipeline(loom::Tensor1 x, loom::Tensor1 y) {
  auto mul = [](float xi, float yi) -> float { return xi * yi; };
  auto reduce = [](loom::Tensor1 t) -> float { return loom::Tensor1::reduce_sum(t); };
  return reduce(loom::Tensor1::map2(mul, x, y));
}

File: experiments/source/python/dot_pipeline.py
import loom


@loom.entry
def dot_pipeline(x: loom.Tensor1, y: loom.Tensor1) -> float:
    def mul(xi: float, yi: float) -> float:
        return xi * yi

    def reduce(t: loom.Tensor1) -> float:
        return loom.Tensor1.reduce_sum(t)

    return reduce(loom.Tensor1.map2(mul, x, y))
